utils/ipfs.py

```diff
@@ -0,0 +1,222 @@
+
+
+class IPFSUploadResult(TypedDict):
+    cid: str
+    text: str
+    messages: list[Tuple[str, str]]
+
+
+# upload text to web3.storage ipfs
+def _upload_str_to_web3_storage(text: str) -> str:
+    text_bytes = text.encode("utf-8")
+    text_file = io.BytesIO(text_bytes)
+    web3_storage_token = get_web3_storage_token()
+
+    endpoint = "https://api.web3.storage/upload"
+    headers = {"Authorization": f"Bearer {web3_storage_token}", "Content-Type": "application/x-directory"}
+
+    response = requests.post(endpoint, headers=headers, data=text_file)
+    response.raise_for_status()
+    response_json = response.json()
+
+    return response_json.get("cid")
+
+
+def _upload_str_to_ipfs(text: str) -> str:
+    return _upload_str_to_web3_storage(text)
+
+
+# calculate cid hash from utf8 str
+def calculate_cid_hash(text: str) -> str:
+    data = bytes(text, "utf-8")
+    return cid_sha256_hash(data)
+
+
+# fetching url status
+async def _fetch_status(session, url: str) -> int:
+    async with session.get(url) as response:
+        return response.status
+
+
+# fetch cid from different api concurrency
+async def _fetch_cid_status_from_ipfs_async(cid: str) -> int:
+    if not cid:
+        return 404
+
+    request_urls = [
+        f"https://{cid}.ipfs.w3s.link",  # faster for uploaded files
+        f"https://api.web3.storage/status/{cid}",  # much faster for not uploaded files
+    ]
+
+    async with aiohttp.ClientSession() as session:
+        tasks = [asyncio.create_task(_fetch_status(session, url)) for url in request_urls]
+        done, pending = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
+
+        for task in pending:
+            task.cancel()
+
+        for task in done:
+            return task.result()
+
+
+def verify_ipfs_description(text: str) -> list[Tuple[str, str]]:
+    messages: list[Tuple[str, str]] = []
+    if not text:
+        messages.append(
+            (
+                "error",
+                (
+                    "You provided an empty string as description. "
+                    "If you provide text as a description, it will help users to make a voting decision. "
+                    "The description is stored on the IPFS network and does not take up much space in the script."
+                ),
+            )
+        )
+
+    address_raw_groups = re.findall(rf"([^`]{REG_ETH_ADDRESS}|{REG_ETH_ADDRESS}[^`])", f" {text} ")
+    if address_raw_groups:
+        address_raw = list(map(lambda x: x[1] or x[2], address_raw_groups))
+        messages.append(
+            (
+                "warning",
+                (
+                    "You have wallet addresses in description which has no Markdown style. "
+                    "You could use inline code block to make it looks better. "
+                    "You need to add '`' before and after the address. Here is the list of addresses:\n"
+                    f"{linesep.join(address_raw)}"
+                ),
+            )
+        )
```

Comment:
:brain: Please enforce all of the ETH addresses to be:
1) Checksummed
2) Their checksum is correct

see: https://eips.ethereum.org/EIPS/eip-55